modules/nextflow/src/main/groovy/nextflow/processor/TaskProcessor.groovy

```diff
@@ -1939,9 +1940,14 @@ class TaskProcessor {
 
             if( item instanceof Path || coerceToPath ) {
                 def path = normalizeToPath(item)
-                def target = executor.isForeignFile(path) ? batch.addToForeign(path) : path
-                def holder = new FileHolder(target)
-                files << holder
+                if (executor.isForeignFile(path)) {
+                    def target = batch.addToForeign(path)
+                    def holder = new RemoteFileHolder(target, path)
+                    files << holder
+                } else {
+                    def holder = new FileHolder(path)
+                    files << holder
+                }
```

Comment:
Why didn't you just do `new FileHolder(path, target)` ? You shouldn't need a new class

Reply:
Sorry, just saw your extra comment

Reply:
However I think my point still stands. You should be able to just do `new FileHolder(path, target)` and add a method to FileHolder to access the source path, e.g.:

```groovy
Path getSourcePath() {
  sourceObj instanceof Path ? sourceObj : null
}
```

Reply:
Np, I don't like the solution of introducing an extra class here either.

I just thought it'd be better to get your opinion first ^^